README.md
# Mopti

![Ruby](https://github.com/yoshoku/mopti/workflows/Ruby/badge.svg)
[![Gem Version](https://badge.fury.io/rb/mopti.svg)](https://badge.fury.io/rb/mopti)
[![BSD 3-Clause License](https://img.shields.io/badge/License-BSD%203--Clause-orange.svg)](https://github.com/yoshoku/mopti/blob/main/LICENSE.txt)
[![Documentation](https://img.shields.io/badge/api-reference-blue.svg)](https://yoshoku.github.io/mopti/doc/)

Mopti is a multivariate optimization library in Ruby.
Mopti supports Nelder-Mead simplex method and Scaled Conjugate Gradient.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'mopti'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install mopti

## Documentation

- [Mopti API Documentation](https://yoshoku.github.io/mopti/doc/)

## Usage

Example 1. Linear Regression with Nelder-Mead simplex method

```ruby
require 'numo/narray'
require 'numo/gnuplot'
require 'mopti'

# Define objective function.
obj_fnc = proc do |w, x, y|
  n_samples = x.shape[0]
  ((y - x.dot(w))**2).sum.fdiv(n_samples)
end

# Explanatory variables.
x = Numo::DFloat[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Dependent variables.
y = 3 * x + 2

# Extend variable for intercept.
xb = Numo::DFloat.vstack([x, [1] * x.size]).transpose.dup

# Optimize parameter vectors.
res = Mopti::minimize(algorithm: 'Nelder-Mead', fnc: obj_fnc, x_init: Numo::DFloat.zeros(2), args: [xb, y])

# Output result.
pp res

a, b = res[:x].to_a
Numo.gnuplot do
  set(terminal: 'png')
  set(output: 'example1.png')
  plot [[0, 10], [a * x[0] + b, a * x[-1] + b], w: :lines, lw: 1, t: 'parameters'],
       [x, y, pt: 6, ps: 2, t: 'data']
end
```

```
$ brew install gnuplot
$ gem install mopti numo-narray numo-gnuplot
$ ruby example1.rb
{:x=>Numo::DFloat(view)#shape=[2]
[3, 2],
 :n_fev=>177,
 :n_iter=>91,
 :fnc=>2.290874014308807e-13}
```

![example1](https://user-images.githubusercontent.com/5562409/74737586-79c9db00-5298-11ea-9063-90e4655f878a.png)


## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/yoshoku/mopti. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/yoshoku/mopti/blob/main/CODE_OF_CONDUCT.md).


## Code of Conduct

Everyone interacting in the Mopti project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/yoshoku/mopti/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
# 0.2.3
- Refactor codes and configs with RuboCop.

# 0.2.2
- Remove dependent gem's type declaration file from installation files.

# 0.2.1
- Fix version specifier of runtime dependencies.

# 0.2.0
- Add type declaration files.
- Refactor some codes with type check.
- Fix some configuration files.

# 0.1.0
- First release.


spec/mopti_spec.rb
# frozen_string_literal: true

RSpec.describe Mopti do
  describe '#minimize' do
    let(:x) { Numo::DFloat.zeros(3) }
    let(:f) { Numo::DFloat[[1, 1, 1], [1, 1, 0], [1, 0, 1], [1, 0, 0], [1, 0, 0]] }
    let(:k) { Numo::DFloat[1.0, 0.3, 0.5] }

    let(:fnc) do
      proc do |x, f, k|
        log_pdot = f.dot(x)
        log_z = Numo::NMath.log(Numo::NMath.exp(log_pdot).sum)
        (log_z - k.dot(x)).to_f
      end
    end

    let(:jcb) do
      proc do |x, f, k|
        log_pdot = f.dot(x)
        log_z = Numo::NMath.log(Numo::NMath.exp(log_pdot).sum)
        p = Numo::NMath.exp(log_pdot - log_z)
        f.transpose.dot(p) - k
      end
    end

    let(:res) { described_class.minimize(algorithm: algorithm, **args) }
    let(:err) { (fnc.call(res[:x], f, k) - fnc.call(Numo::DFloat[0, -0.524869316, 0.487525860], f, k)).abs }

    shared_examples 'perform optimization' do
      it 'obtains optimized vector that minimizes the value of the objective function' do
        expect(err).to be < 1.0e-6
      end
    end

    context 'when optimizer is scaled conjugate gradient' do
      let(:algorithm) { 'SCG' }
      let(:args) { { fnc: fnc, jcb: jcb, x_init: x, args: [f, k] } }

      it_behaves_like 'perform optimization'
    end

    context 'when optimizer is Nelder-Mead simplex' do
      let(:algorithm) { 'Nelder-Mead' }
      let(:args) { { fnc: fnc, x_init: x, args: [f, k] } }

      it_behaves_like 'perform optimization'
    end

    context 'when non-existent optimizer is given' do
      it 'raises ArgumentError' do
        expect { described_class.minimize(algorithm: 'foobar', fnc: fnc) }.to raise_error(ArgumentError)
      end
    end
  end
end


spec/spec_helper.rb
# frozen_string_literal: true

require 'bundler/setup'
require 'mopti'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/mopti/nelder_mead_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe Mopti::NelderMead do
  let(:x) { Numo::DFloat.zeros(3) }
  let(:f) { Numo::DFloat[[1, 1, 1], [1, 1, 0], [1, 0, 1], [1, 0, 0], [1, 0, 0]] }
  let(:k) { Numo::DFloat[1.0, 0.3, 0.5] }

  let(:fnc) do
    proc do |x, f, k|
      log_pdot = f.dot(x)
      log_z = Numo::NMath.log(Numo::NMath.exp(log_pdot).sum)
      (log_z - k.dot(x)).to_f
    end
  end

  let(:optimizer) { described_class.new(fnc: fnc, x_init: x, args: [f, k]) }
  let(:params) { optimizer.map { |p| p } }
  let(:res) { params.last }
  let(:err) { (fnc.call(res[:x], f, k) - fnc.call(Numo::DFloat[0, -0.524869316, 0.487525860], f, k)).abs }

  it { expect(params.size).to be_positive }
  it { expect(res.keys).to match(%i[x n_fev n_iter fnc]) }
  it { expect(err).to be < 1.0e-6 }
end


lib/mopti.rb
# frozen_string_literal: true

require 'numo/narray'

require 'mopti/version'
require 'mopti/nelder_mead'
require 'mopti/scaled_conjugate_gradient'

# Mopti is a multivariate optimization library in Ruby.
module Mopti
  module_function

  # Perform minization of the objective function.
  #
  # @param algorithm [String] Type of optimizer.
  #   - 'SCG': ScaledConjugateGradient
  #   - 'Nelder-Mead': NelderMead
  # @return [Hash] Result of optimization.
  def minimize(algorithm:, **args)
    optimizer = case algorithm
                when 'SCG'
                  ScaledConjugateGradient.new(**args)
                when 'Nelder-Mead'
                  NelderMead.new(**args)
                else
                  raise ArgumentError, 'A non-existent algorithm is specified'
                end
    res = nil
    optimizer.each { |params| res = params }
    res
  end
end


lib/mopti/scaled_conjugate_gradient.rb
# frozen_string_literal: true

module Mopti
  # ScaledConjugateGradient is a class that implements multivariate optimization using scaled conjugate gradient method.
  #
  # @example
  #   # Seek the minimum value of the expression a*u**2 + b*u*v + c*v**2 + d*u + e*v + f for
  #   # given values of the parameters and an initial guess (u, v) = (0, 0).
  #   # https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.fmin_cg.html#scipy.optimize.fmin_cg
  #   require 'numo/narray'
  #   require 'mopti'
  #
  #   args = [2, 3, 7, 8, 9, 10]
  #
  #   f = proc do |x, a, b, c, d, e, f|
  #     u = x[0]
  #     v = x[1]
  #     a * u**2 + b * u * v + c * v**2 + d * u + e * v + f
  #   end
  #
  #   g = proc do |x, a, b, c, d, e, f|
  #     u = x[0]
  #     v = x[1]
  #     gu = 2 * a * u + b * v + d
  #     gv = b * u + 2 * c * v + e
  #     Numo::DFloat[gu, gv]
  #   end
  #
  #   x0 = Numo::DFloat.zeros(2)
  #
  #   optimizer = Mopti::ScaledConjugateGradient.new(fnc: f, jcb: g, x_init: x0, args: args)
  #   result = optimizer.map { |params| params }.last
  #
  #   pp result
  #
  #   # {:x=>Numo::DFloat#shape=[2]
  #   # [-1.80847, -0.25533],
  #   #  :n_fev=>10,
  #   #  :n_jev=>18,
  #   #  :n_iter=>9,
  #   #  :fnc=>1.6170212789006122,
  #   #  :jcb=>1.8698188678645777e-07}
  #
  # *Reference*
  # 1. Moller, M F., "A Scaled Conjugate Gradient Algorithm for Fast Supervised Learning," Neural Networks, Vol. 6, pp. 525--533, 1993.
  class ScaledConjugateGradient
    include Enumerable

    # Create a new optimizer with scaled conjugate gradient.
    #
    # @param fnc [Method/Proc] Method for calculating the objective function to be minimized.
    # @param jcb [Method/Proc] Method for calculating the gradient vector.
    # @param args [Array/Hash] Arguments pass to the 'fnc' and 'jcb'.
    # @param x_init [Numo::NArray] Initial point.
    # @param max_iter [Integer] Maximum number of iterations.
    # @param xtol [Float] Tolerance for termination for the optimal vector norm.
    # @param ftol [Float] Tolerance for termination for the objective function value.
    # @param jtol [Float] Tolerance for termination for the gradient norm.
    def initialize(fnc:, jcb:, x_init:, args: nil, max_iter: 200, xtol: 1e-6, ftol: 1e-8, jtol: 1e-7)
      @fnc = fnc
      @jcb = jcb
      @x_init = x_init
      @args = args
      @max_iter = max_iter
      @ftol = ftol
      @jtol = jtol
      @xtol = xtol
    end

    # Iteration for optimization.
    #
    # @overload each(&block) -> Object
    # @yield [Hash] { x:, n_fev:, n_jev:, n_iter:, fnc:, jcb: }
    #   - x [Numo::NArray] Updated vector by optimization.
    #   - n_fev [Interger] Number of calls of the objective function.
    #   - n_jev [Integer] Number of calls of the jacobian.
    #   - n_iter [Integer] Number of iterations.
    #   - fnc [Float] Value of the objective function.
    #   - jcb [Numo::Narray] Values of the jacobian
    # @return [Enumerator] If block is not given, this method returns Enumerator.
    def each
      return to_enum(__method__) unless block_given?

      x = @x_init
      f_prev = func(x, @args)
      n_fev = 1
      f_curr = f_prev
      j_next = jacb(x, @args)
      n_jev = 1

      j_curr = j_next.dot(j_next)
      j_prev = j_next.dup
      d = -j_next
      success = true
      n_successes = 0
      beta = 1.0

      n_iter = 0

      while n_iter < @max_iter
        if success
          mu = d.dot(j_next)
          if mu >= 0.0
            d = -j_next
            mu = d.dot(j_next)
          end
          kappa = d.dot(d)
          break if kappa < 1e-16

          sigma = SIGMA_INIT / Math.sqrt(kappa)
          x_plus = x + (sigma * d)
          j_plus = jacb(x_plus, @args)
          n_jev += 1
          theta = d.dot(j_plus - j_next) / sigma
        end

        delta = theta + (beta * kappa)
        if delta <= 0
          delta = beta * kappa
          # TODO: Investigate the cause of the type error.
          # Cannot assign a value of type `::Complex` to a variable of type `::Float`
          # beta -= theta / kappa
          beta = (beta - (theta / kappa)).to_f
        end
        alpha = -mu / delta

        x_next = x + (alpha * d)
        f_next = func(x_next, @args)
        n_fev += 1

        delta = 2 * (f_next - f_prev) / (alpha * mu)
        if delta >= 0
          success = true
          n_successes += 1
          x = x_next
          f_curr = f_next
        else
          success = false
          f_curr = f_prev
        end

        n_iter += 1
        yield({ x: x, n_fev: n_fev, n_jev: n_jev, n_iter: n_iter, fnc: f_curr, jcb: j_curr })

        if success
          break if (f_next - f_prev).abs < @ftol
          break if (alpha * d).abs.max < @xtol

          f_prev = f_next

          j_prev = j_next
          j_next = jacb(x, @args)
          n_jev += 1

          j_curr = j_next.dot(j_next)
          break if j_curr <= @jtol
        end

        beta = beta * 4 < BETA_MAX ? beta * 4 : BETA_MAX if delta < 0.25
        beta = beta / 4 > BETA_MIN ? beta / 4 : BETA_MIN if delta > 0.75

        if n_successes == x.size
          d = -j_next
          beta = 1.0
          n_successes = 0
        elsif success
          gamma = (j_prev - j_next).dot(j_next) / mu
          d = -j_next + (gamma * d)
        end
      end
    end

    SIGMA_INIT = 1e-4
    BETA_MIN = 1e-15
    BETA_MAX = 1e+15

    private_constant :SIGMA_INIT, :BETA_MIN, :BETA_MAX

    private

    def func(x, args)
      if args.is_a?(Hash)
        @fnc.call(x, **args)
      elsif args.is_a?(Array)
        @fnc.call(x, *args)
      elsif args.nil?
        @fnc.call(x)
      else
        @fnc.call(x, args)
      end
    end

    def jacb(x, args)
      if args.is_a?(Hash)
        @jcb.call(x, **args)
      elsif args.is_a?(Array)
        @jcb.call(x, *args)
      elsif args.nil?
        @jcb.call(x)
      else
        @jcb.call(x, args)
      end
    end
  end
end


lib/mopti/nelder_mead.rb
# frozen_string_literal: true

module Mopti
  # NelderMead is a class that implements multivariate optimization using the Nelder-Mead simplex method.
  #
  # @example
  #   require 'numo/narray'
  #   require 'mopti'
  #
  #   args = Numo::DFloat[2, 3]
  #
  #   f = proc { |x, a| (8 * (x - a)**2).sum }
  #
  #   x0 = Numo::DFloat.zeros(2)
  #
  #   optimizer = Mopti::NelderMead.new(fnc: f, x_init: x0, args: args)
  #   result = optimizer.map { |params| params }.last
  #
  #   pp result
  #
  #   # {:x=>Numo::DFloat(view)#shape=[2]
  #   # [2, 3],
  #   #  :n_fev=>165,
  #   #  :n_iter=>84,
  #   #  :fnc=>5.694864987422661e-13}
  #
  # *Reference*
  # 1. Gao, F. and Han, L., "Implementing the Nelder-Mead simplex algorithm with adaptive parameters," Computational Optimization and Applications, vol. 51 (1), pp. 259--277, 2012.
  class NelderMead
    include Enumerable

    # Create a new optimizer with the Nelder-Mead simplex method.
    #
    # @param fnc [Method/Proc] Method for calculating the objective function to be minimized.
    # @param args [Array/Hash] Arguments pass to the 'fnc' and 'jcb'.
    # @param x_init [Numo::NArray] Initial point.
    # @param max_iter [Integer] Maximum number of iterations.
    #   If nil is given, max_iter sets to 200 * number of dimensions.
    # @param xtol [Float] Tolerance for termination for the optimal vector norm.
    # @param ftol [Float] Tolerance for termination for the objective function value.
    def initialize(fnc:, x_init:, args: nil, max_iter: nil, xtol: 1e-6, ftol: 1e-6)
      @fnc = fnc
      @args = args
      @x_init = x_init
      @max_iter = max_iter
      @xtol = xtol
      @ftol = ftol
    end

    # Iteration for optimization.
    #
    # @overload each(&block) -> Object
    # @yield [Hash] { x:, n_fev:, n_jev:, n_iter:, fnc:, jcb: }
    #   - x [Numo::NArray] Updated vector by optimization.
    #   - n_fev [Interger] Number of calls of the objective function.
    #   - n_iter [Integer] Number of iterations.
    #   - fnc [Float] Value of the objective function.
    # @return [Enumerator] If block is not given, this method returns Enumerator.
    def each
      return to_enum(__method__) unless block_given?

      x = @x_init.dup
      n = x.size
      max_iter = @max_iter || (200 * n)

      alpha = 1.0
      beta = n > 1 ? 1 + 2.fdiv(n) : 2.0
      gamma = n > 1 ? 0.75 - 1.fdiv(2 * n) : 0.5
      delta = n > 1 ? 1 - 1.fdiv(n) : 0.5

      sim = x.class.zeros(n + 1, n)
      sim[0, true] = x
      n.times do |k|
        y = x.dup
        y[k] = (y[k]).zero? ? ZERO_TAU : (1 + NON_ZERO_TAU) * y[k]
        sim[k + 1, true] = y
      end

      fsim = Numo::DFloat.zeros(n + 1)

      (n + 1).times { |k| fsim[k] = func(sim[k, true], @args) }
      n_fev = n + 1

      ind = fsim.sort_index
      fsim = fsim[ind].dup
      sim = sim[ind, true].dup

      n_iter = 0
      while n_iter < max_iter
        break if ((sim[1..-1, true] - sim[0, true]).abs.flatten.max <= @xtol) && ((fsim[0] - fsim[1..-1]).abs.max <= @ftol)

        xbar = sim[0...-1, true].sum(0) / n
        xr = xbar + (alpha * (xbar - sim[-1, true]))
        fr = func(xr, @args)
        n_fev += 1

        shrink = true
        if fr < fsim[0]
          xe = xbar + (beta * (xr - xbar))
          fe = func(xe, @args)
          n_fev += 1
          shrink = false
          if fe < fr
            sim[-1, true] = xe
            fsim[-1] = fe
          else
            sim[-1, true] = xr
            fsim[-1] = fr
          end
        elsif fr < fsim[-2]
          shrink = false
          sim[-1, true] = xr
          fsim[-1] = fr
        elsif fr < fsim[-1]
          xoc = xbar + (gamma * (xr - xbar))
          foc = func(xoc, @args)
          n_fev += 1
          if foc <= fr
            shrink = false
            sim[-1, true] = xoc
            fsim[-1] = foc
          end
        else
          xic = xbar - (gamma * (xr - xbar))
          fic = func(xic, @args)
          n_fev += 1
          if fic < fsim[-1]
            shrink = false
            sim[-1, true] = xic
            fsim[-1] = fic
          end
        end

        if shrink
          (1..n).to_a.each do |j|
            sim[j, true] = sim[0, true] + (delta * (sim[j, true] - sim[0, true]))
            fsim[j] = func(sim[j, true], @args)
            n_fev += 1
          end
        end

        ind = fsim.sort_index
        sim = sim[ind, true].dup
        fsim = fsim[ind].dup

        n_iter += 1

        yield({ x: sim[0, true], n_fev: n_fev, n_iter: n_iter, fnc: fsim.min })
      end
    end

    NON_ZERO_TAU = 0.05
    ZERO_TAU = 0.00025

    private_constant :NON_ZERO_TAU, :ZERO_TAU

    private

    def func(x, args)
      if args.is_a?(Hash)
        @fnc.call(x, **args)
      elsif args.is_a?(Array)
        @fnc.call(x, *args)
      elsif args.nil?
        @fnc.call(x)
      else
        @fnc.call(x, args)
      end
    end
  end
end
